pkg/controller/clusterdeployment/clusterdeployment_controller.go

```diff
@@ -44,7 +45,19 @@ import (
 )
 
 // controllerKind contains the schema.GroupVersionKind for this controller type.
-var controllerKind = hivev1.SchemeGroupVersion.WithKind("ClusterDeployment")
+var (
+	controllerKind                    = hivev1.SchemeGroupVersion.WithKind("ClusterDeployment")
+	ownedByClusterDeploymentListTypes = []runtime.Object{
+		&batchv1.JobList{},
+		&corev1.PersistentVolumeClaimList{},
+		&corev1.SecretList{},
+		&hivev1.ClusterProvisionList{},
+		&hivev1.ClusterDeprovisionList{},
+		&hivev1.DNSZoneList{},
+		&hivev1.ClusterStateList{},
```

Comment:
The controller that creates the object should be the one responsible for ensuring the owner reference.

Reply:
Is that always going to be possible?

What about a situation where a controller creates the object, but doesn't watch the owner object?

What will reconcile it then?

Also, is this a change from what we discussed, or did I just not understand you before?

I thought you were saying that the owner of the object should reconcile the owned objects.

Reply:
> Is that always going to be possible?
>
> What about a situation where a controller creates the object, but doesn't watch the owner object?

What would prompt a controller to create an object on behalf of owner that the controller is not reconciling? I cannot think of a scenario. How would a controller know to create an object for the owner in that case?

> What will reconcile it then?
> 
> Also, is this a change from what we discussed, or did I just not understand you before?
> 
> I thought you were saying that the owner of the object should reconcile the owned objects.

This is what I recall us discussing. I don't know what it means for "the owner of the object [to] reconcile". Objects do no reconcile: Controllers reconcile. Typically, there is a single controller that reconciles a particular object. However, Hive has multiple controllers that reconcile the same object, presumably in order to modularize features. If a controller is creating an object as part of a feature that is isolated to that controller, then it breaks modularity to leak the responsibility of maintaining the owner relationship of that object to a different controller.

Reply:
Interesting, I must have misunderstood you.

What you're proposing sounds a lot like what I was initially proposing (the controller of the owned object reconciles the ownership), but I thought you suggested to do it this way.

Let's discuss tomorrow at stand up.

Reply:
It seems like there is still a misunderstanding. I am not suggesting that the controller reconciling the owned object does anything. The controller that created the object should be responsible for maintaining the owner relationship. The controller that created the object should not be the controller that reconciles the object. For example, the clusterDeployment controller reconciles clusterDeployment resources. That controller creates clusterProvision  resources owned by clusterDeployment resources. The clusterDeployment controller should maintain the owner relationship between the clusterProvision resource and the clusterDeployment resource.

Reply:
Let's look at another example where the clusterDeployment controller should not be maintaining the relationship even though the owner is a ClusterDeployment resource. The clusterState controller reconciles ClusterDeployment resources and creates ClusterState resources owned by ClusterDeployment resources. The clusterState controller should be the controller responsible for maintaining the owner relationship between ClusterState resources and ClusterDeployment resources.